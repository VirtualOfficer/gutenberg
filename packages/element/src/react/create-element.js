/**
 * External dependencies
 */
import { createElement } from 'react';

/**
 * Returns a new element of given type. Type can be either a string tag name or
 * another function which itself returns an element.
 *
 * @param {?(string|Function)} type     Tag name or element creator.
 * @param {Object}             props    Element properties, either attribute
 *                                       set to apply to node or values to
 *                                       pass through to element creator.
 * @param {...WPElement}       children Descendant elements.
 *
 * @return {WPElement} Element.
 */
export default createElement;
