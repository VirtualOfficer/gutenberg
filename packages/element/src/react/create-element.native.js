/**
 * External dependencies
 */
import { get, isString } from 'lodash';
import { createElement as createElementDefault } from 'react';
import {
	Circle,
	Defs,
	Ellipse,
	G,
	Line,
	Path,
	Polygon,
	TextPath,
	TSpan,
	Svg,
} from 'react-native-svg';

const htmlTagToNativeComponent = {
	circle: Circle,
	defs: Defs,
	ellipse: Ellipse,
	g: G,
	line: Line,
	path: Path,
	polygon: Polygon,
	svg: Svg,
	textPath: TextPath,
	tspan: TSpan,
	/*Polyline,
	Text,
	Rect,
	Use,
	Image,
	Symbol,
	LinearGradient,
	RadialGradient,
	Stop,
	ClipPath*/
};

const mapHTMLTagToNativeComponent = ( type ) => {
	if ( ! isString( type ) ) {
		return type;
	}

	return get( htmlTagToNativeComponent, [ type ], type );
};

/**
 * Returns a new element of given type. Type can be either a string tag name or
 * another function which itself returns an element.
 *
 * @param {?(string|Function)} type     Tag name or element creator.
 * @param {Object}             props    Element properties, either attribute
 *                                       set to apply to node or values to
 *                                       pass through to element creator.
 * @param {...WPElement}       children Descendant elements.
 *
 * @return {WPElement} Element.
 */
const createElement = ( type, props, ...children ) => {
	return createElementDefault(
		mapHTMLTagToNativeComponent( type ),
		props,
		...children
	);
};

export default createElement;
