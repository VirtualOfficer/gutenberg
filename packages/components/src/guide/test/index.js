/**
 * External dependencies
 */
import { fireEvent, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

/**
 * Internal dependencies
 */
import Guide from '../';

describe( 'Guide', () => {
	it( 'renders nothing when there are no pages', () => {
		render( <Guide pages={ [] } /> );
		expect( screen.queryByRole( 'dialog' ) ).not.toBeInTheDocument();
	} );

	it( 'renders one page at a time', () => {
		render(
			<Guide
				pages={ [
					{ content: <p>Page 1</p> },
					{ content: <p>Page 2</p> },
				] }
			/>
		);

		expect( screen.queryByRole( 'dialog' ) ).toBeVisible();
		expect( screen.queryByText( 'Page 1' ) ).toBeVisible();
		expect( screen.queryByText( 'Page 2' ) ).not.toBeInTheDocument();
	} );

	it( 'hides back button and shows forward button on the first page', () => {
		render(
			<Guide
				pages={ [
					{ content: <p>Page 1</p> },
					{ content: <p>Page 2</p> },
				] }
			/>
		);

		expect(
			screen.queryByRole( 'button', { name: 'Previous' } )
		).not.toBeInTheDocument();
		expect(
			screen.queryByRole( 'button', { name: 'Next' } )
		).toBeVisible();
		expect(
			screen.queryByRole( 'button', { name: 'Finish' } )
		).not.toBeInTheDocument();
	} );

	it( 'shows back button and shows finish button on the last page', async () => {
		const user = userEvent.setup( {
			advanceTimers: jest.advanceTimersByTime,
		} );
		render(
			<Guide
				pages={ [
					{ content: <p>Page 1</p> },
					{ content: <p>Page 2</p> },
				] }
			/>
		);
		await user.click( screen.getByRole( 'button', { name: 'Next' } ) );

		expect( screen.queryByText( 'Page 1' ) ).not.toBeInTheDocument();
		expect( screen.queryByText( 'Page 2' ) ).toBeVisible();

		expect(
			screen.queryByRole( 'button', { name: 'Previous' } )
		).toBeVisible();
		expect(
			screen.queryByRole( 'button', { name: 'Next' } )
		).not.toBeInTheDocument();
		expect(
			screen.queryByRole( 'button', { name: 'Finish' } )
		).toBeVisible();
	} );

	it( "doesn't display the page control if there is only one page", () => {
		render( <Guide pages={ [ { content: <p>Page 1</p> } ] } /> );
		expect(
			screen.queryByRole( 'list', { name: 'Guide controls' } )
		).not.toBeInTheDocument();
	} );

	it( 'calls onFinish when the finish button is clicked', async () => {
		const user = userEvent.setup( {
			advanceTimers: jest.advanceTimersByTime,
		} );
		const onFinish = jest.fn();
		render(
			<Guide
				onFinish={ onFinish }
				pages={ [ { content: <p>Page 1</p> } ] }
			/>
		);
		await user.click( screen.getByRole( 'button', { name: 'Finish' } ) );

		expect( onFinish ).toHaveBeenCalled();
	} );

	it( 'calls onFinish when the modal is closed', () => {
		const onFinish = jest.fn();
		render(
			<Guide
				onFinish={ onFinish }
				pages={ [ { content: <p>Page 1</p> } ] }
			/>
		);

		/**
		 * Workaround to trigger an Escape keypress event.
		 *
		 * @todo Remove this workaround in favor of userEvent.keyboard() or userEvent.type().
		 *
		 * This curently doesn't work:
		 *
		 * ```
		 * await user.keyboard( '[Escape]' );
		 * ```
		 *
		 * because the event sent has a `keyCode` of `0`.
		 *
		 * To fix this, we'll need to update the Modal component to work with `KeyboardEvent.code`.
		 *
		 * @see https://github.com/testing-library/user-event/issues/969
		 */
		fireEvent.keyDown( screen.getByRole( 'dialog' ), {
			key: 'Escape',
			keyCode: 27,
		} );

		expect( onFinish ).toHaveBeenCalled();
	} );
} );
