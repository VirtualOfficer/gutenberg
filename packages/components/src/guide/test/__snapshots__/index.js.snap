// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Guide renders when there are pages 1`] = `
<WithInstanceId(Modal)
  className="components-guide"
  shouldCloseOnClickOutside={false}
>
  <KeyboardShortcuts
    key="0"
    shortcuts={
      Object {
        "left": [Function],
        "right": [Function],
      }
    }
  />
  <div
    className="components-guide__container"
  >
    <GuidePage>
      Page 1
    </GuidePage>
    <div
      className="components-guide__footer"
    >
      <PageControl
        currentPage={0}
        numberOfPages={2}
        setCurrentPage={[Function]}
      />
      <ForwardRef(IconButton)
        className="components-guide__forward-button"
        icon={<ForwardButtonIcon />}
        onClick={[Function]}
      >
        Next
      </ForwardRef(IconButton)>
    </div>
  </div>
</WithInstanceId(Modal)>
`;
